taichi/transforms/auto_diff.cpp

```diff
@@ -1140,6 +1375,122 @@ class MakeAdjoint : public ADTransform {
     accumulate(stmt->val, insert<GlobalLoadStmt>(adjoint_ptr));
     stmt->parent->erase(stmt);
   }
+
+  void visit(MatrixPtrStmt *stmt) override {
+    if (stmt->origin->is<GlobalPtrStmt>() ||
+        stmt->origin->is<ExternalPtrStmt>()) {
+      // Handled in GlobalLoadStmt and GlobalStoreStmt
+      return;
+    }
+
+    Stmt *adjoint_value = nullptr;
+    if (stmt->offset->is<ConstStmt>()) {
```

Comment:
Shall we move the adjoint creation into `Stmt *adjoint(Stmt *stmt)` ? Maybe we can only handle the logic of chain rule here.

Reply:
On a second thought, I feel like it's actually working on the chain rule instead of the `adjoint` (I can change the name a little bit).

Taking the following example, say we have a MatrixPtrStmt in the forward pass:

```
f32* $1 = MatrixPtrStmt(Tensor<4 x f32>* $origin, i32 $offset)
```

The problem is that there's shape mismatch between the adjoint of `$origin` (`Tensor<4 x f32>*`) and the adjoint of `$1` (`f32*`). The `adjoint_value` here denotes the adjoint for `$origin` and the entire effort here is to construct the `Tensor<4 x f32>` adjoint from `f32*` adjoint.

So it's actually working on the chain rule for MatrixPtrStmt